llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp

```diff
@@ -660,14 +671,16 @@ void PPCAsmPrinter::EmitTlsCall(const MachineInstr *MI,
          "GETtls[ld]ADDR[32] must read GPR3");
 
   if (Subtarget->isAIXABI()) {
-    // On AIX, the variable offset should already be in R4 and the region handle
-    // should already be in R3.
-    // For TLSGD, which currently is the only supported access model, we only
-    // need to generate an absolute branch to .__tls_get_addr.
+    // For TLSGD, the variable offset should already be in R4 and the region
+    // handle should already be in R3. We generate an absolute branch to
+    // .__tls_get_addr. For TLSLD, the module handle should already be in R3.
+    // We generate an absolute branch to .__tls_get_mod.
     Register VarOffsetReg = Subtarget->isPPC64() ? PPC::X4 : PPC::R4;
     (void)VarOffsetReg;
-    assert(MI->getOperand(2).isReg() &&
-           MI->getOperand(2).getReg() == VarOffsetReg &&
+    assert((MI->getOpcode() == PPC::GETtlsMOD32AIX ||
+            MI->getOpcode() == PPC::GETtlsMOD64AIX ||
+            (MI->getOperand(2).isReg() &&
+             MI->getOperand(2).getReg() == VarOffsetReg)) &&
            "GETtls[ld]ADDR[32] must read GPR4");
     EmitAIXTlsCallHelper(MI);
```

Comment:
The helper functions have special calling convention properties. For example, they do not use the FP registers. The IBM XL compiler was able to take advantage of that.

For:
```c
__attribute__((tls_model("local-dynamic"))) __thread int x;
double g(int, double);
void f() {
  double gg = g(0, 1.);
  g(x, gg);
}
```

The IBM XL compilers were able to make use of the returned `double` staying in the register:
```
      28: 48 00 00 03   bla 0
                        00000028:  R_RBA        (idx: 36) .__tls_get_mod[PR]
      2c: 7c 66 18 2e   lwzx 3, 6, 3
      30: 4b ff ff d1   bl 0x0 <.f>
                        00000030:  R_RBR        (idx: 34) .g[PR]
```

Clang/LLVM loads the value from the stack:
```
      24: 48 00 00 03   bla 0
                0000000000000024:  R_RBA        (idx: 3) .__tls_get_mod[PR]
      28: e8 82 00 08   ld 4, 8(2)
                000000000000002a:  R_TOC        (idx: 17) x[TC]
      2c: 7c 63 22 aa   lwax 3, 3, 4
      30: 4b ff ff d1   bl 0x0 <.f>
                0000000000000030:  R_RBR        (idx: 1) .g[PR]
```

Reply:
From what I've observed, Clang/LLVM uses the special GPR usage rules for the call to __tls_get_mod.  But as you noted, FPRs are not handled as a special case.

Reply:
Thank you for looking into this!

I tried the example, and it seems turn on optimization can help remove the FP load. We declared "Defs = [X0,X4,X5,X11,LR8,CR0]" for GETtlsMOD64AIX, and "Defs = [R0,R4,R5,R11,LR,CR0]" for GETtlsMOD32AIX. I think those FP registers should be treated as not touched by the call to __tls_get_mod.

Reply:
Thanks @orcguru. I misread the assembly 🤦‍♂️ (the extra instruction is not an `lfd`).